openedx/core/djangoapps/content_libraries/api.py

```diff
@@ -412,13 +412,32 @@ def get_library_team(library_key):
     ]
 
 
+def get_library_user_permissions(library_key, user):
+    """
+    Fetch the specified user's access information.
+    """
+    ref = ContentLibrary.objects.get_by_key(library_key)
+    grant = ref.permission_grants.filter(user=user).first()
+    if grant is None:
+        return
```

Comment:
A few reasons:

1. It would allow email lookup by username for any user who has studio-library access, if this object is permitted to be returned by the view, which another client writer would probably do without thinking about what data it would leak.
2. The type disparation offers some better guarantee that a successful return won't be treated as 'read' access by a developer that isn't thinking or forgot to check for a specific constant to be present in the grant.
3. We lose the ability to treat the response in a boolean fashion.

While the above could be mitigated by having the client views do a check for if grant.access_level == NO_ACCESS, it's quite a bit more wordy and not more clear, in my opinion, for the code to say 'This person's access level is NO_ACCESS' rather than to just say 'No permission granted'.

I'm also not sure what other None checks you're referring to-- this code is only called presently in views.py and set_library_user_permissions within this file, and in all instances that I can see would still require a (more verbose) conditional statement to handle the case of no access having been granted.

Reply:
@Kelketek 

> 1. It would allow email lookup by username for any user who has studio-library access, if this object is permitted to be returned by the view, which another client writer would probably do without thinking about what data it would leak.

Good point. Thank you for considering that.

> 2. The type disparation offers some better guarantee that a successful return won't be treated as 'read' access by a developer that isn't thinking or forgot to check for a specific constant to be present in the grant.

Possibly. A foggy-headed developer could also treat a 'read' grant as a 'write' grant if they forget to check that access constant.

> 3. We lose the ability to treat the response in a boolean fashion.

> While the above could be mitigated by having the client views do a check for if grant.access_level == NO_ACCESS, it's quite a bit more wordy and not more clear, in my opinion, for the code to say 'This person's access level is NO_ACCESS' rather than to just say 'No permission granted'.

This is a good point. However, it begs a further question for me: why return an attrs object at all, and not just an access-level string or None? The other data in the object is redunant, as `.user` always equals `user` and `.group` always equals None. It seems that the object could be instantiated on-the-fly in the one place it's passed to a serializer. 

Since it's just a question of style, I'll leave it up to you whether to leave or change it. Either way, please do update the docstring to be explicit about the fact that None is meaningful return value from this function.

Reply:
> This is a good point. However, it begs a further question for me: why return an attrs object at all, and not just an access-level string or None? The other data in the object is redunant, as .user always equals user and .group always equals None. It seems that the object could be instantiated on-the-fly in the one place it's passed to a serializer.

I'm not sure what the group field is for. If we don't intend to use it then you're probably right that this could stand to be a simpler function. Was it part of the spec?

Either way, for now I've updated the docstring to say that None is a useful return, and to further make it clear, I've made `return None` explicit.

Reply:
> I'm not sure what the group field is for.

@Kelketek I'm not sure whether you don't why (1) the `.group` field exists on `ContentLibraryPermission`, or (2) the `.group` field is returned from this function. I'll assume both, so bear with me if I end up explaining stuff you already know :)

1. The [`ContentLibraryPermission.group`](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/content_libraries/models.py#L89-L91) field, when supplied, provides the permission to every user in the group. So a user can have access to a library via (a) a direct permission grant, (b) a membership in a group with a permission grant, or (c) both. The mechanics behind this are seem to be orchestrated through rules in `permissions.py`. If this is new to you, I recommend getting familiar with that system, and then giving your PR a sanity check. 

2. For `get_library_user_permissions`, it seems that this function only considers user-level permissions and ignores group-level ones (although it would be good to make that clear in the docstring as well). We mandate (at a Python level) that [the `.user` and `.group` fields are mutually exclusive](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/content_libraries/models.py#L89), so it is safe to assume that `.group` will always be `None` in this function. This is why I think it's safe to simplify the function's return value.

> Was it part of the spec?

As far as I'm aware, there are no specs coming from the edX side of what this Python API needs to look like (other than a general request of "expose and use a good Python API" :)

Reply:
I put the `group` functionality in - at the time I thought it was something we'd need for LabXchange, though in the end we're not using it. It's a simple and common use case of [plain old Django groups](https://docs.djangoproject.com/en/3.1/topics/auth/default/#groups) though, and could be very handy. Say you have a video team, and the video team maintains 50+ video libraries on edx.org. Whenever someone joins or leaves the video team, you don't want to have to manually add/remove that user to every single video library. Instead, you give the "video team group" admin permissions on every library, then you can just add/remove users from the video team. Much more convenient to administer. Groups can be created to represent teams, organizations, or anything.

I actually wrote a working [REST API for administering groups](https://github.com/edx/edx-platform/pull/22423) but never merged it since we weren't using it.

Reply:
@kdmccormick Talked this over with @bradenmacdonald and I think I'm going to leave this function as-is. While it is true that it might be a bit more than is needed for the task, I'd like to keep functions like this one using the same datatypes. Especially since we tend to feed them right into serializers. If we didn't do this I'd either have to return just a string from the endpoint (which seems a bit odd to me), or specify the data for the serializer manually.

As it currently is, we get to pass the value right over to the serializer and return it, keeping the view code succinct.